src/lib/comms/sol-lwm2m.c

```diff
@@ -2822,19 +2822,25 @@ handle_read(struct sol_lwm2m_client *client,
         }
     }
 
-    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
-        r = setup_tlv(res, &buf);
-        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
-        sol_lwm2m_resource_clear(res);
-    }
-
     r = add_coap_int_option(resp, SOL_COAP_OPTION_CONTENT_FORMAT,
         &format, sizeof(format));
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
     r = set_packet_payload(resp, (const uint8_t *)buf.data, buf.used);
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
+    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
+        r = setup_tlv(res, &buf);
+        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+    }
+
+    /* Go through the list another time to prevent double free in case of
+     * an error in setup_tlv (which triggers free for all resources in the
+     * error path processing). */
+    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
```

Comment:
Well, the problem is that `goto err_exit` leads to another freeing of the data from all items in the vector. If it was already freed half way, it would lead to double free for the already freed items. Anyway, I think a simpler fix is just setting the freed pointer to `NULL`, because after that the error path can call `free()` as many times it likes. Will update the patch now...